### ISSUE TITLE
Provide an option to turn off buildserver detection magic

### ISSUE THREAD

**Issue by dittert** [GitTools/GitVersion#791]
I would like to suggest a command line parameter to turn off build server detection completely. In our setup `GitVersion.exe` is called from some unit tests on repositories that do not have any connection to the repository the build is running from. 

As a consequence those unit tests run on developer machines but fail on TeamCity:

```
INFO [02/21/16 13:35:24:30] Working directory: D:\TeamCity\buildAgent\temp\buildTmp\junit7400125167335433572
INFO [02/21/16 13:35:24:31] Applicable build agent found: 'TeamCity'.
INFO [02/21/16 13:35:24:33] Project root is: D:\TeamCity\buildAgent\temp\buildTmp\junit7400125167335433572
INFO [02/21/16 13:35:24:40] Begin: Loading version variables from disk cache
INFO [02/21/16 13:35:24:42] Cache file D:\TeamCity\buildAgent\temp\buildTmp\junit7400125167335433572\.git\gitversion_cache\44DCD4383285104839031850AAF389949188A60D.yml not found.
INFO [02/21/16 13:35:24:42] End: Loading version variables from disk cache (Took: 18.00ms)
WARN [02/21/16 13:35:24:42] TeamCity doesn't make the current branch available through environmental variables.

Depending on your authentication and transport setup of your git VCS root things may work. In that case, ignore this warning.

In your TeamCity build configuration, add a parameter called `system.Git_Branch` with value %teamcity.build.vcs.branch.<vcsid>%

See http://gitversion.readthedocs.org/en/latest/build-server-support/build-server/teamcity for more info
INFO [02/21/16 13:35:24:43] Branch from build environment: 
WARN [02/21/16 13:35:24:44] An error occurred:
0 remote(s) have been detected. When being run on a build server, the Git repository is expected to bear one (and no more than one) remote.
```

In this case, there is temporary Git repository that only contains local branches with `master` being checked out. I'm wondering what `GitVersion.exe` needs the remote for (as `/nofetch`) was specified.

**Comment by JakeGinnivan** [comment]
This warning can probably be removed when /nofetch is specified.

The first one though is so we can detect if we are building in a PR without a fetch.

**Comment by pvandervelde** [comment]
Has there been any progress on this @JakeGinnivan? I'm running into an issue where I perform a merge on the Jenkins build server (e.g. merge feature/myfeature into develop, ending up on develop) and then running GitVersion. This however completely changes my Git history. 

My history before running GitVersion looks like:

    *   2356fd9 (HEAD -> develop) Merge branch 'feature/builds' into develop
      |\  
      | * a8e6f9e (origin/feature/builds, feature/builds) Invalid path for the deployment
      | * ec23484 testing builds
      | * 6592ffe Deploy target for the zip file
      | * 639a29a Updating the nuspec to match the new project names
      | * 7c926c8 Removing the suo file because it's user specific
      | * 43276b5 Adding the nuspec
      | * 18a2037 Fixing compiler warnings
      | * ed14c33 Updating to reflect the new location of the packages folder
      | * 41de87f Adding Nuclei.Build so that we can embed version control information in the assembly
      | * fab66b3 Updating solution with the new solution configuration
      | * 951a344 Linking the projects to the base configuration file
      | * e6be8b7 Adding the base properties for each project and the rulesets.
      | * 30741e7 Adding the standard build scripts
      |/  
      * 8621d57 (origin/develop) Updated startup project
      *   0666ef1 Merge branch 'feature/Changing_unit_testing_to_mstest' into develop
      |\  
      | * 159eea4 Deleted unnecessary nuget dependency and reference to nuget targets
      | * 280dc89 Updated project references
      | * 3ebaecb Renamed folders to reflect namespaces
      | * b21338f Removed another reference to xunit
      | * fa6a95a Removed references to xunit
      | * c71398c Converted from xunit to vstest
      |/  
      *   81fdcd8 Merge branch 'feature/DependencyWalkerIntegration' into develop
      |\  
      | * 1e49023 Reordered files to put properties after constructors
      | * 8b8c2b1 Now supports when a project references both a nuget and an assembly with the same name
      | * b6e1a01 Renamed references in test project file
      | * c7f203c Added comments to test cases Deleted files that are not needed
      | * ff881a6 Added comments to test case project files Deleted unnecessary files
      | * 9e968bc Cleaned up test case directory
      | * 51deac6 Fixed things in code review
      | * 5fa32a4 Added basic tests for ReferenceHandler
      | * e101ce1 -Added tests for AssemblyClash - Split code in ReferenceHandler to smaller methods - Added App.config to walker tests to make them not run in parallel, previously it was messing with the static nodes list
      | * 4988c43 Removed unnecessary code Added test cases for dependency walker Made some classes static Fixed bug where nuget nodes would overwrite others in walker
      | * 85a24fd Removed unnecessary code
      | * 9137774 Reimplemented recursive project reading
      | * 06fbce3 Refactored walker: - Fixed bug where nuget nodes would overwrite others - Simplified reading a project
      | * a766252 Added integration tests from original Dependency Walker
      | * d7f470e Fixed nuget dependency version bug
      | * bf92250 Fixed issue where a reference to an assembly which has not been created in memory yet was skipped sometimes
      | * d3a624a Cleaned up xml doc Removed unnecessary files
      | * ef3e156 Changed string creation to use string builder Removed unnecessary constructor from TextDependencyVerifierFileWriter Removed .close() where using statements are used
      | * f4c8f6c Added walker xml output
      | * 69f5736 Updated text output to be more friendly for walker
      | * b136f39 Cleaned up dependency walker code
      | * 2b4e1f7 Added dependency walker core
      |/  
      * f0c41a8 Fixed text file output
      * 28254ed Added warnings to text output
      * bca66df Updated XML output
      * fb75fcc Changed some variable declarations to use var
      * 6451e7a Updated XML output Moved formatting to text and xml formatter
      * f15e7ab Remove unnecessary whitespace
      * c262c1e Moved some formatting functionality out of ReferenceHandler
      * 720e91f Added null checks
      *   fddbf85 Merge branch 'feature/Better_Structure' into develop
      |\  
      | * 364549d Updated to have a relative path
      | * 4423b27 Changed System.IO.Path to Path
      | * 5bdd82d Updated the structure to use a dictionary
      |/  
      * 34c613f Deleted packages
      * 4c45fbe Removed packages
      * 6ec11dc Updated for correct output
      * 05bf88f Updated missing files
      * 067cc25 Added missing files
      *   a539b69 Merge branch 'feature/outputimprovements' into develop
      |\  
      | * 49beb03 Added method input verification
      | * 9062d23 Updated assembly loading
      | * 8e68794 Simplified GetClashes()
      | * 8b0c7fe Updated GacFilesReferenced to not be static Removed ProcessReferences from ReferenceHandler
      | * da7aa4c Made directory reader more accurate Fixed typo in test case name
      | * 870b7e4 Simplified directory reader
      | * cb46c20 Added directory reader tests
      | * 6455961 Log can now be configured from the command line
      | * 3e89057 Added initial logging
      | * 18b25eb Added ReferenceHandler tests and test cases
      | * 380ce13 Updated default namespace
      | * 72cceaa Updated text and xml writer to have using statements
      | * 02d66c6 Renamed CheckIfAllReferencesAccessible to ProcessReferences
      | * 08de923 Fixed bug where it ignored files in the root directory
      | * 7ee271c Changed some explicit types to var
      | * 518bd55 Updated notation
      | * 3330c9c Updated console app project name
      | * 436ad71 Moved output to IReferenceFormatter Updated the console app to use Mono.Options
      | * fe53fb8 Cleaned up and renamed most things to make it easier to understand Created xml documentation
      | * aae2075 Updated git ignore
      | * 3a355ec Continued cleaning up re code review
      | * db9eeb0 Deleted files that should have been deleted before Started changes from code review
      | * 7f6ade0 moved processing out of output methods
      | * a077e69 Updated gitignore
      | * 16166be Changed output to text file
      | * 19c0454 Refactored a lot to make it simpler
      | * d6be2a5 Made output easier to read
      | * 5094209 Fixed output text format Added exclude directory options Added list of dlls used from GAC Added test cases for reference file Refactored processReferences and DirectoryVersionClashes
      | * 211a027 Added unit tests for ReferenceFile
      | * 0fa6cd7 Made all DLLs work, not just the first of them
      | * 711c379 Cleaned up naming conventions
      | * 765c23d Checks dll and exe references now Renamed classes accordingly Sorted clashing dll output alphabetically
      * | 8487990 Deleted /src/DependencyVerifierClassLibraryTestCases
      |/  
      *   e4aa73a (origin/master) Merge branch 'feature/ConsoleApplicationStructure' into develop
      |\  
      | * 5f2e342 Console app structure improved - started test cases -updated to check dependencies from dll file
      * | 6fbba3b no message
      |/  
      * 7f874dd Added everything to date - Can traverse directories and show what exes depend on dlls - Can show directories where two versions of a dll are expected (bad D: )
      * 4d2bd6b Added file Readme.md

The history after running git version looks like:

    *   2356fd9 (HEAD -> feature/builds) Merge branch 'feature/builds' into develop
      |\  
      | * a8e6f9e (origin/feature/builds) Invalid path for the deployment
      | * ec23484 testing builds
      | * 6592ffe Deploy target for the zip file
      | * 639a29a Updating the nuspec to match the new project names
      | * 7c926c8 Removing the suo file because it's user specific
      | * 43276b5 Adding the nuspec
      | * 18a2037 Fixing compiler warnings
      | * ed14c33 Updating to reflect the new location of the packages folder
      | * 41de87f Adding Nuclei.Build so that we can embed version control information in the assembly
      | * fab66b3 Updating solution with the new solution configuration
      | * 951a344 Linking the projects to the base configuration file
      | * e6be8b7 Adding the base properties for each project and the rulesets.
      | * 30741e7 Adding the standard build scripts
      |/  
      * 8621d57 (origin/develop, develop) Updated startup project
      *   0666ef1 Merge branch 'feature/Changing_unit_testing_to_mstest' into develop
      |\  
      | * 159eea4 Deleted unnecessary nuget dependency and reference to nuget targets
      | * 280dc89 Updated project references
      | * 3ebaecb Renamed folders to reflect namespaces
      | * b21338f Removed another reference to xunit
      | * fa6a95a Removed references to xunit
      | * c71398c Converted from xunit to vstest
      |/  
      *   81fdcd8 Merge branch 'feature/DependencyWalkerIntegration' into develop
      |\  
      | * 1e49023 Reordered files to put properties after constructors
      | * 8b8c2b1 Now supports when a project references both a nuget and an assembly with the same name
      | * b6e1a01 Renamed references in test project file
      | * c7f203c Added comments to test cases Deleted files that are not needed
      | * ff881a6 Added comments to test case project files Deleted unnecessary files
      | * 9e968bc Cleaned up test case directory
      | * 51deac6 Fixed things in code review
      | * 5fa32a4 Added basic tests for ReferenceHandler
      | * e101ce1 -Added tests for AssemblyClash - Split code in ReferenceHandler to smaller methods - Added App.config to walker tests to make them not run in parallel, previously it was messing with the static nodes list
      | * 4988c43 Removed unnecessary code Added test cases for dependency walker Made some classes static Fixed bug where nuget nodes would overwrite others in walker
      | * 85a24fd Removed unnecessary code
      | * 9137774 Reimplemented recursive project reading
      | * 06fbce3 Refactored walker: - Fixed bug where nuget nodes would overwrite others - Simplified reading a project
      | * a766252 Added integration tests from original Dependency Walker
      | * d7f470e Fixed nuget dependency version bug
      | * bf92250 Fixed issue where a reference to an assembly which has not been created in memory yet was skipped sometimes
      | * d3a624a Cleaned up xml doc Removed unnecessary files
      | * ef3e156 Changed string creation to use string builder Removed unnecessary constructor from TextDependencyVerifierFileWriter Removed .close() where using statements are used
      | * f4c8f6c Added walker xml output
      | * 69f5736 Updated text output to be more friendly for walker
      | * b136f39 Cleaned up dependency walker code
      | * 2b4e1f7 Added dependency walker core
      |/  
      * f0c41a8 Fixed text file output
      * 28254ed Added warnings to text output
      * bca66df Updated XML output
      * fb75fcc Changed some variable declarations to use var
      * 6451e7a Updated XML output Moved formatting to text and xml formatter
      * f15e7ab Remove unnecessary whitespace
      * c262c1e Moved some formatting functionality out of ReferenceHandler
      * 720e91f Added null checks
      *   fddbf85 Merge branch 'feature/Better_Structure' into develop
      |\  
      | * 364549d Updated to have a relative path
      | * 4423b27 Changed System.IO.Path to Path
      | * 5bdd82d Updated the structure to use a dictionary
      |/  
      * 34c613f Deleted packages
      * 4c45fbe Removed packages
      * 6ec11dc Updated for correct output
      * 05bf88f Updated missing files
      * 067cc25 Added missing files
      *   a539b69 Merge branch 'feature/outputimprovements' into develop
      |\  
      | * 49beb03 Added method input verification
      | * 9062d23 Updated assembly loading
      | * 8e68794 Simplified GetClashes()
      | * 8b0c7fe Updated GacFilesReferenced to not be static Removed ProcessReferences from ReferenceHandler
      | * da7aa4c Made directory reader more accurate Fixed typo in test case name
      | * 870b7e4 Simplified directory reader
      | * cb46c20 Added directory reader tests
      | * 6455961 Log can now be configured from the command line
      | * 3e89057 Added initial logging
      | * 18b25eb Added ReferenceHandler tests and test cases
      | * 380ce13 Updated default namespace
      | * 72cceaa Updated text and xml writer to have using statements
      | * 02d66c6 Renamed CheckIfAllReferencesAccessible to ProcessReferences
      | * 08de923 Fixed bug where it ignored files in the root directory
      | * 7ee271c Changed some explicit types to var
      | * 518bd55 Updated notation
      | * 3330c9c Updated console app project name
      | * 436ad71 Moved output to IReferenceFormatter Updated the console app to use Mono.Options
      | * fe53fb8 Cleaned up and renamed most things to make it easier to understand Created xml documentation
      | * aae2075 Updated git ignore
      | * 3a355ec Continued cleaning up re code review
      | * db9eeb0 Deleted files that should have been deleted before Started changes from code review
      | * 7f6ade0 moved processing out of output methods
      | * a077e69 Updated gitignore
      | * 16166be Changed output to text file
      | * 19c0454 Refactored a lot to make it simpler
      | * d6be2a5 Made output easier to read
      | * 5094209 Fixed output text format Added exclude directory options Added list of dlls used from GAC Added test cases for reference file Refactored processReferences and DirectoryVersionClashes
      | * 211a027 Added unit tests for ReferenceFile
      | * 0fa6cd7 Made all DLLs work, not just the first of them
      | * 711c379 Cleaned up naming conventions
      | * 765c23d Checks dll and exe references now Renamed classes accordingly Sorted clashing dll output alphabetically
      * | 8487990 Deleted /src/DependencyVerifierClassLibraryTestCases
      |/  
      *   e4aa73a (origin/master, master) Merge branch 'feature/ConsoleApplicationStructure' into develop
      |\  
      | * 5f2e342 Console app structure improved - started test cases -updated to check dependencies from dll file
      * | 6fbba3b no message
      |/  
      * 7f874dd Added everything to date - Can traverse directories and show what exes depend on dlls - Can show directories where two versions of a dll are expected (bad D: )

Notice that in the first case the develop branch is the current branch and feature/builds has been merged into develop. In the second case develop is back where it was before the merge and feature/builds is suddenly the active branch on the merge commit.

The GitVersion log looks like:
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:09] Working directory: c:\ci\workspace\Tools\DependencyVerifier_Feature\packages\nBuildKit.MsBuild.0.8.2\build
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:10] IsDynamicGitRepository: False
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:12] Returning Project Root from DotGitDirectory: c:\ci\workspace\Tools\DependencyVerifier_Feature\.git - c:\ci\workspace\Tools\DependencyVerifier_Feature
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:12] Running on Windows.
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:13] Applicable build agent found: 'Jenkins'.
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:13] Branch from build environment: feature/builds
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:31] IsDynamicGitRepository: False
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:31] Returning Project Root from DotGitDirectory: c:\ci\workspace\Tools\DependencyVerifier_Feature\.git - c:\ci\workspace\Tools\DependencyVerifier_Feature
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:31] Project root is: c:\ci\workspace\Tools\DependencyVerifier_Feature
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:31] DotGit directory is: c:\ci\workspace\Tools\DependencyVerifier_Feature\.git
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:32] IsDynamicGitRepository: False
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:32] Returning Project Root from DotGitDirectory: c:\ci\workspace\Tools\DependencyVerifier_Feature\.git - c:\ci\workspace\Tools\DependencyVerifier_Feature
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:33] Begin: Loading version variables from disk cache
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:33] Cache file c:\ci\workspace\Tools\DependencyVerifier_Feature\.git\gitversion_cache\1EBEC90C15502806EB419BA3CDB9B25DEDC2BCFD.yml not found.
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:33] End: Loading version variables from disk cache (Took: 1.00ms)
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:33] IsDynamicGitRepository: False
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:33] Returning Project Root from DotGitDirectory: c:\ci\workspace\Tools\DependencyVerifier_Feature\.git - c:\ci\workspace\Tools\DependencyVerifier_Feature
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:34] Using latest commit on specified branch
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:35] Begin: Attempting to inherit branch configuration from parent branch
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:36] HEAD is merge commit, this is likely a pull request using develop as base
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:37] Begin: Finding branch source
    2016-10-17 17:48:15		    INFO [10/17/16 17:48:15:39] Begin: Found merge base of e4aa73a02b574d9c8a1d22b1a7448b8b97e8f67d against master
    2016-10-17 17:48:15		    INFO [10/17/16 17:48:15:40] End: Found merge base of e4aa73a02b574d9c8a1d22b1a7448b8b97e8f67d against master (Took: 11.00ms)
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:40] End: Finding branch source (Took: 29.99ms)
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:41] Begin: Getting branches containing the commit 'e4aa73a02b574d9c8a1d22b1a7448b8b97e8f67d'.
    2016-10-17 17:48:15		    INFO [10/17/16 17:48:15:41] Trying to find direct branches.
    2016-10-17 17:48:15		    INFO [10/17/16 17:48:15:41] Direct branch found: 'master'.
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:41] End: Getting branches containing the commit 'e4aa73a02b574d9c8a1d22b1a7448b8b97e8f67d'. (Took: 4.00ms)
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:41] Found possible parent branches: master
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:41] End: Attempting to inherit branch configuration from parent branch (Took: 64.01ms)
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:42] Running against branch: feature/builds (2356fd93db4674cdf84e0f2a7c5f12df604421a9)
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:42] Begin: Calculating base versions
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:48] Fallback base version: 0.1.0 with commit count source 4d2bd6b7b647e41f0c0d77844326d437a7ebdfbd
    2016-10-17 17:48:15		  INFO [10/17/16 17:48:15:62] Base version used: Fallback base version: 0.1.0 with commit count source 4d2bd6b7b647e41f0c0d77844326d437a7ebdfbd
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:62] End: Calculating base versions (Took: 199.99ms)
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:63] Skipping version increment
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:63] Using branch name to calculate version tag
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:65] 100 commits found between 4d2bd6b7b647e41f0c0d77844326d437a7ebdfbd and 2356fd93db4674cdf84e0f2a7c5f12df604421a9
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:66] Begin: Creating dictionary
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:67] End: Creating dictionary (Took: 3.99ms)
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:67] Begin: Storing version variables to cache file c:\ci\workspace\Tools\DependencyVerifier_Feature\.git\gitversion_cache\1EBEC90C15502806EB419BA3CDB9B25DEDC2BCFD.yml
    2016-10-17 17:48:15		INFO [10/17/16 17:48:15:72] End: Storing version variables to cache file c:\ci\workspace\Tools\DependencyVerifier_Feature\.git\gitversion_cache\1EBEC90C15502806EB419BA3CDB9B25DEDC2BCFD.yml (Took: 48.01ms)

Is there anything I can do about this? It's very much destroying my 'cunning' plan for testing merges during the build.

**Comment by pvandervelde** [comment]
@JakeGinnivan Is there any chance this could go into 4.0?
